bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,467 @@
+    """Contains a symbol and the BeautifulSoup object needed to parse it."""
+
+    symbol: DocItem
+    soup: BeautifulSoup
+
+    def __eq__(self, other: Union[QueueItem, DocItem]):
+        if isinstance(other, DocItem):
+            return self.symbol == other
+        return NamedTuple.__eq__(self, other)
+
+
+class CachedParser:
+    """
+    Get symbol markdown from pages with smarter caching.
+
+    DocItems are added through the `add_item` method which adds them to the `_page_symbols` dict.
+    `get_markdown` is used to fetch the markdown; when this is used for the first time on a page,
+    all of the symbols are queued to be parsed to avoid multiple web requests to the same page.
+    """
+
+    def __init__(self):
+        self._queue: List[QueueItem] = []
+        self._results = {}
+        self._page_symbols: Dict[str, List[DocItem]] = defaultdict(list)
+        self._item_events: Dict[DocItem, asyncio.Event] = {}
+        self._parse_task = None
+
+    async def get_markdown(self, client_session: ClientSession, doc_item: DocItem) -> str:
+        """
+        Get result markdown of `doc_item`.
+
+        If no symbols were fetched from `doc_item`s page before,
+        the HTML has to be fetched before parsing can be queued.
+        """
+        if (symbol := self._results.get(doc_item)) is not None:
+            return symbol
+
+        if (symbols_to_queue := self._page_symbols.get(doc_item.url)) is not None:
+            async with client_session.get(doc_item.url) as response:
+                soup = BeautifulSoup(await response.text(encoding="utf8"), "lxml")
+
+            self._queue.extend(QueueItem(symbol, soup) for symbol in symbols_to_queue)
+            del self._page_symbols[doc_item.url]
+            log.debug(f"Added symbols from {doc_item.url} to parse queue.")
+
+            if self._parse_task is None:
+                self._parse_task = asyncio.create_task(self._parse_queue())
+
+        self._move_to_front(doc_item)
+        self._item_events[doc_item] = item_event = asyncio.Event()
+        await item_event.wait()
+        return self._results[doc_item]
+
+    async def _parse_queue(self) -> None:
+        """
+        Parse all item from the queue, setting associated events for symbols if present.
+
+        The coroutine will run as long as the queue is not empty, resetting `self._parse_task` to None when finished.
+        """
+        log.trace("Starting queue parsing.")
+        while self._queue:
+            item, soup = self._queue.pop()
+            try:
+                markdown = get_symbol_markdown(soup, item)
+                await doc_cache.set_if_exists(item, markdown)
+                self._results[item] = markdown
+            except Exception:
+                log.exception(f"Unexpected error when handling {item}")
+            else:
+                if (event := self._item_events.get(item)) is not None:
+                    event.set()
+            await asyncio.sleep(0.1)
```

Comment:
This is used to allow the bot to do other stuff during the fully synchronous parsing while keeping the load a bit lighter since it's not a priority.
Because we await above we could remove this, causing the bot to take all available resources and maybe (?) slowing down some other parts of it, but the whole queue would be parsed considerably faster
It could also be moved into a thread but I felt like it'd complicate the code a bit too much while this is sufficient

I'm not too sure on how python async would like the parsing consuming everything available while only giving other parts of the control during the redis IO waits or the devops side of things; should it be removed or is this kind of a "sleep to do other things" await fine

Reply:
I would have used a thread cause that's what I'm familiar with, but if this works then I suppose it's fine. I don't know enough to comment on it.

Reply:
I hadn't really considered it till now, but ran some tests now and the markdown parsing which is fully synchronous took up to 250ms after I fetched some pages, I suppose this would be an issue for the real bot. I'll look into switching it over whole to a thread and if that's not viable without too many complications at least the parsing itself can run in an executor

Reply:
Tried putting the whole parsing task into a thread but it was a bit of hard to wrap around all the locking after I was done, while still being unsure if there aren't any more race conditions around. Decided to go with the executor because of  that and I'll leave this as it is now unless someone else has an opinion against it